qemu/tests/numa_cpu.py

```diff
@@ -0,0 +1,132 @@
+                die_min, die_max = _get_boundary(die, vcpu_dies, die_weight)
+                core_min, core_max = _get_boundary(core, vcpu_cores, core_weight)
+                thread_min, thread_max = _get_boundary(thread, vcpu_threads, thread_weight)
+                cpu_min = socket_min + die_min + core_min + thread_min
+                cpu_max = socket_max + die_max + core_max + thread_max
+                if nodeid in tmp.keys():
+                    tmp[nodeid] += range(cpu_min, cpu_max+1, 1)
+                else:
+                    tmp[nodeid] = range(cpu_min, cpu_max+1, 1)
+            elif vm_arch in ('ppc64le', 'ppc64'):
+                if nodeid in tmp.keys():
+                    tmp[nodeid] += [int(core)]
+                else:
+                    tmp[nodeid] = [int(core)]
+        for item in sorted(tmp.items(), key=lambda item:item[0]):
+            expected_cpus.append(set(sorted(item[1])))
+        return expected_cpus
+
+    def get_hotpluggable_cpus():
+        """
+        Parse output of "query-hotpluggable-cpus", get the cpus that bind to
+        special node.
+        """
+        hotpluggable_cpus = []
+        tmp = {}
+        out = vm.monitor.info("hotpluggable-cpus")
+        for vcpu_info in out:
+            vcpu_info = vcpu_info["props"]
+            nodeid = vcpu_info.get("node-id")
+            if nodeid is not None:
+                socketid = vcpu_info["socket-id"]
+                dieid = vcpu_info["die-id"]
+                coreid = vcpu_info["core-id"]
+                threadid = vcpu_info["thread-id"]
+                cpu_index = socketid * socket_weight + dieid * die_weight \
+                            + coreid * core_weight + threadid * thread_weight
+                if nodeid in tmp.keys():
+                    tmp[nodeid] += [int(cpu_index)]
+                else:
+                    tmp[nodeid] = [int(cpu_index)]
+        for item in sorted(tmp.items(), key=lambda item:item[0]):
+            hotpluggable_cpus.append(set(sorted(item[1])))
+        return hotpluggable_cpus
+
+    vm = env.get_vm(params["main_vm"])
+    session = vm.wait_for_login()
+    os_type = params["os_type"]
+    vm_arch = params["vm_arch_name"]
+
+    # vcpu_sockets = vm.cpuinfo.sockets
+    # vcpu_dies = vm.cpuinfo.dies
+    # vcpu_cores = vm.cpuinfo.cores
+    # vcpu_threads = vm.cpuinfo.threads
+    vcpu_sockets = params.get_numeric('vcpu_sockets')
+    vcpu_dies = params.get_numeric('vcpu_dies')
+    vcpu_cores = params.get_numeric('vcpu_cores')
+    vcpu_threads = params.get_numeric('vcpu_threads')
+
+    socket_weight = vcpu_dies * vcpu_cores * vcpu_threads
+    die_weight = vcpu_cores * vcpu_threads
+    core_weight = vcpu_threads
+    thread_weight = 1
+
+    hotpluggable_cpus = get_hotpluggable_cpus()
+    logging.info("Hotpluggable cpus: %s" % hotpluggable_cpus)
+    numa_cpu_monitor = [item[1] for item in vm.monitors[0].info_numa()]
```

Comment:
Can we modify `vm.monitors[0]` to the main monitor `vm.monitor`?

Reply:
> Can we modify `vm.monitors[0]` to the main monitor `vm.monitor`?

Good catch, use main monitor is better, thanks!